### PR TITLE
Modifications for Off-Heap Technology for Large Arrays

### PR DESCRIPTION
**Pull request by LinHu2016**
Modifications required to complete the introduction of an optimized
memory management scheme for large arrays in a region-based 
garbage collector.
More specifically, this is a new heap scheme and large
object organization (in-heap and off-heap), an
enhanced method for allocating large objects (proxy
objects), and an optimized technique of accessing
large arrays during runtime (JIT optimizations).

 - new define OMR_GC_SPARSE_HEAP_ALLOCATION
 - set OMR_GC_SPARSE_HEAP_ALLOCATION ON
 - obsolete Trc_MM_double_map_Entry, once all usages are updated.
 - replace J9VM_GC_ENABLE_SPARSE_HEAP_ALLOCATION with
   OMR_GC_SPARSE_HEAP_ALLOCATION.
 - empty method offsetOfContiguousDataAddrField() in ObjectModel
 - empty method generateArrayElementShiftAmountTrees() in TransformUtil
 - empty method isOffHeapAllocationEnabled() in ObjectModel

Signed-off-by: hulin <linhu@ca.ibm.com>

### CODE REVIEW

**Comment by LinHu2016** [comment]
@amicic @dmitripivkine please review the changes, Thanks

File: gc/base/GCExtensionsBase.hpp [hunk 1]
```diff
@@ -1509,12 +1517,12 @@ class MM_GCExtensionsBase : public MM_BaseVirtual {
 		, largePageWarnOnError(false)
 		, largePageFailOnError(false)
 		, largePageFailedToSatisfy(false)
-#if defined(OMR_GC_DOUBLE_MAP_ARRAYLETS)
 		, isArrayletDoubleMapRequested(false)
 		, isArrayletDoubleMapAvailable(false)
-#endif /* defined(OMR_GC_DOUBLE_MAP_ARRAYLETS) */
```
Comment (dmitripivkine) [inline gc/base/GCExtensionsBase.hpp:1515]: These two fields are defined under the same flag. If `if defined` is removed here, it should be removed at fields definitions above as well.

Comment (amicic) [reply gc/base/GCExtensionsBase.hpp:1515]: good observation. a preference is to keep the ifdefs if possible (if there are no spots in any current or more importantly new code that uses the variables without being wrapped with the same build flag)

File: gc/base/j9mm.tdf [hunk 2]
```diff
@@ -1028,3 +1028,5 @@ TraceEvent=Trc_MM_MSSSS_flip_restore_tilt_after_percolate_current_status Overhea
 
 TraceEvent=Trc_MM_clearheap Overhead=1 Level=1 Group=gclogger Template="clearheap with free memory size: %zu, object size: %zu"
 TraceEvent=Trc_MM_cleanUpSegmentsInAnonymousClassLoader_stats Overhead=1 Level=1 Group=gclogger Template="Anonymous classes: total: %zu, unloaded: %zu, searched ROM: %zu, found ROM fast: %zu, found ROM slow: %zu, not found ROM: %zu"
+
+TraceEntry=Trc_MM_double_map_EntryNew Overhead=1 Level=3 Group=arraylet Template="MM_IndexableObjectAllocationModel::doubleMapArraylets. originalDataSize: %p, adjustedDataSize: %p, spine: %p, leafSize: %p leavesCount: %zu"
```
Comment (amicic) [inline gc/base/j9mm.tdf:1032]: we have to remember to obsolete the old one, once all (downstream) usages are updated

File: gc/base/GCExtensionsBase.cpp [hunk 3]
```diff
@@ -356,3 +357,9 @@ MM_GCExtensionsBase::usingSATBBarrier()
 #endif /* defined(OMR_GC_REALTIME) */
 	return false;
 }
+
+bool
+MM_GCExtensionsBase::isAddressWithinHeap(void *address)
```
Comment (amicic) [inline gc/base/GCExtensionsBase.cpp:362]: let's try without introducing this method. instead, we can do:
- simplest from testing perspective, but probably least desirable, introduce it on J9 side
- use (NULL != regionManager->regionDescriptorForAddress())  instead, which probably also a reasonable simple compromise
- examine the caller sites more carefully and see if it's really need at all, or if we can use isDataAddr(Non)Adjacent instead

File: gc/base/GCExtensionsBase.hpp [hunk 4]
```diff
@@ -1512,7 +1520,7 @@ class MM_GCExtensionsBase : public MM_BaseVirtual {
 #if defined(OMR_GC_DOUBLE_MAP_ARRAYLETS)
 		, isArrayletDoubleMapRequested(false)
 		, isArrayletDoubleMapAvailable(false)
-#endif /* defined(OMR_GC_DOUBLE_MAP_ARRAYLETS) */
+#endif /* OMR_GC_DOUBLE_MAP_ARRAYLETS */
```
Comment (amicic) [inline gc/base/GCExtensionsBase.hpp:1523]: removed 'defined()', let's restore the line to what exactly it was

**Comment by 0xdaryl** [comment]
Jenkins build all

**Comment by 0xdaryl** [comment]
@LinHu2016 : I see you've pushed recently.  Does this require a GC re-review or are you just addressing Alek's comments?

**Comment by amicic** [comment]
No further changes are required. 
I believe, the latest push was just a rebase to align with some changes from J9 repo for testing purposes (of J9).

**Review by 0xdaryl — APPROVED** [review]

**Pull request merged**